go-controller/pkg/ovndbmanager/ovndbmanager_test.go

```diff
@@ -132,3 +618,31 @@ func TestElectionTimer(t *testing.T) {
 func keyForArgs(args ...string) string {
 	return strings.Join(args, "-")
 }
+
+// create file with name, fail on error or if the file already exists
+func createDbFile(t *testing.T, name string) {
+	_, err := os.Stat(name)
+	if os.IsNotExist(err) {
+		f, err := os.OpenFile(name, os.O_RDONLY|os.O_CREATE, 0644)
+		if err != nil {
+			t.Errorf("Unexpected errror: %v", err)
+		}
+		f.Close()
+	} else {
+		t.Errorf("File already exists: %s, %v", name, err)
+	}
+}
+
+// failOnErrorMismatch fails either if an error is seen but not expected
+// or if an error is expected but when the substring does not match the error
+func failOnErrorMismatch(t *testing.T, err error, errorString string) {
+	if err != nil {
+		if errorString == "" || !strings.Contains(err.Error(), errorString) {
+			t.Errorf("Unexpected outcome. Got error '%v' but it does not contain expected string '%s'", err, errorString)
+		}
+	} else {
+		if errorString != "" {
+			t.Errorf("No error expected. However, received '%v'", err)
```

Comment:
this looks to me more like we expected an error (`errorString != ""`) but got none (`err == nil`), right?

Reply:
Yes. I wanted to get rid of some duplicate logic inside the test methods.

E.g. the test case looks like this:
~~~
		{
			desc: "Test error: failed to change election timer when leader timer must change but desired is more than double",
			mockCalls: map[string]*mockRes{
				keyForArgs("cluster/change-election-timer", "OVN_Northbound", "3000"): {
					res:    "",
					stderr: "failure",
					err:    fmt.Errorf("failure"),
				},
			},
			timeout:      5000,
			role:         "leader",
			currentTimer: "1500",
			errorString:  "failed to change election timer for",  // <---------- expected err string
		},
~~~

And the call to failOnErrorMismatch looks like this:
~~~
		// fail either if an error is seen but not expected
			// or if an error is expected but when the substring does not match the error
			failOnErrorMismatch(t, err, tc.errorString)
~~~

And method failOnErrorMismatch:
~~~
// failOnErrorMismatch fails either if an error is seen but not expected
// or if an error is expected but when the substring does not match the error
func failOnErrorMismatch(t *testing.T, err error, errorString string) {
	if err != nil {
		if errorString == "" || !strings.Contains(err.Error(), errorString) {
			t.Errorf("Unexpected outcome. Got error '%v' but it does not contain expected string '%s'", err, errorString)
		}
	} else {
		if errorString != "" {
			t.Errorf("No error expected. However, received '%v'", err)	
		}
	}
}
~~~

So we fail on a mismatch between err (returned from the function under test) and errorString (provided in the test cases).

So in other words, if err == nil and errorString != "", then this is invalid.
if err != nil and if errorString == "" or if the errorString does not match the errors Error message, then this is invalid, too.
 
I might rewrite the `} else { if errorString != "" {` but I think the current version is easier on the eyes.

WDYT?

Reply:
It's very well done, only the error message that I was quoting - if I followed everything correctly - should instead be reversed: "Expected error xyz, but got none", right?

Reply:
Indeed. I changed this to:
~~~
// failOnErrorMismatch fails either if an error is seen but not expected
// or if an error is expected but when the substring does not match the error
func failOnErrorMismatch(t *testing.T, receivedErr error, expectedErrorString string) {
	if receivedErr != nil {
		if expectedErrorString == "" {
			t.Errorf("No error expected. However, received '%v' from method under test.", receivedErr)
		} else if !strings.Contains(receivedErr.Error(), expectedErrorString) {
			t.Errorf("Expected error string to contain '%s'. However, method under test threw error '%v'.", expectedErrorString, receivedErr)
		}
	} else if expectedErrorString != "" {
		t.Errorf("Error with error string '%s' expected. However, method under test completed without error.", expectedErrorString)
	}
}
~~~

I also noticed that existing error messages:
~~~
			return fmt.Errorf("failed to change election timer for %s %v %v", db.dbName, err, stderr)
(...)
                        return fmt.Errorf("failed to change election timer for %s %v %v", db.dbName, err, stderr)
~~~

would return "failed to change election timeout for ... failure failure". I changed that to include commas, e.g.:
~~~
return fmt.Errorf("failed to change election timer for %s, %v, %v", db.dbName, err, stderr)
~~~

So that it's a bit clearer that one is err, and the other one stderr.